src/v/transform/transform_manager.cc

```diff
@@ -0,0 +1,386 @@
+// complicated table of indexes.
+class processor_table {
+    struct key_index_t {};
+    struct ntp_index_t {};
+    struct id_index_t {};
+
+    class entry_t {
+    public:
+        entry_t(
+          std::unique_ptr<transform::processor> processor,
+          ss::lw_shared_ptr<transform::probe> probe)
+          : _processor(std::move(processor))
+          , _probe(std::move(probe))
+          , _backoff(std::make_unique<processor_backoff>()) {}
+
+        processor* processor() const { return _processor.get(); }
+        ss::lw_shared_ptr<probe> probe() const { return _probe; }
+        processor_backoff* backoff() const { return _backoff.get(); }
+
+        const model::ntp& ntp() const { return _processor->ntp(); }
+        model::transform_id id() const { return _processor->id(); }
+
+    private:
+        std::unique_ptr<transform::processor> _processor;
+        ss::lw_shared_ptr<transform::probe> _probe;
+        std::unique_ptr<processor_backoff> _backoff;
+    };
+
+    using underlying_t = boost::multi_index::multi_index_container<
+      entry_t,
+      boost::multi_index::indexed_by<
+        // uniquely indexed by id and ntp together
+        boost::multi_index::hashed_unique<
+          boost::multi_index::tag<key_index_t>,
+          boost::multi_index::composite_key<
+            entry_t,
+            boost::multi_index::
+              const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+            boost::multi_index::
+              const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>>,
+          boost::multi_index::composite_key_hash<
+            std::hash<model::transform_id>,
+            std::hash<model::ntp>>>,
+        // indexed by ntp
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<ntp_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, const model::ntp&, &entry_t::ntp>,
+          std::hash<model::ntp>,
+          std::equal_to<>>,
+        // indexed by id
+        boost::multi_index::hashed_non_unique<
+          boost::multi_index::tag<id_index_t>,
+          boost::multi_index::
+            const_mem_fun<entry_t, model::transform_id, &entry_t::id>,
+          std::hash<model::transform_id>,
+          std::equal_to<>>>>;
+
+public:
+    const entry_t& insert(
+      std::unique_ptr<processor> processor,
+      ss::lw_shared_ptr<transform::probe> probe) {
+        auto [it, inserted] = _underlying.emplace(
+          std::move(processor), std::move(probe));
+        vassert(inserted, "invalid transform processor management");
+        return *it;
+    }
+
+    ss::lw_shared_ptr<probe> get_or_create_probe(
+      model::transform_id id, const model::transform_metadata& meta) {
+        auto& id_index = _underlying.get<id_index_t>();
+        auto id_it = id_index.find(id);
+        if (id_it == id_index.end()) {
+            auto probe = ss::make_lw_shared<transform::probe>();
```

Comment:
i'm curious about the lifetime of the probe since it is a shared pointer here. will it outlive the underlying transform? i ask because we've been bit in the past with similar setups where--translating into transform vernacular--the transform would removed and then a new transform would be created with a name that conflicted with metrics that hadn't yet been deregistered because something was holding onto the shared pointer.

Reply:
Right now the probe is only saved alongside transforms, but I think breaking out the boost::multi_index will it so we can probably easily use a unique_ptr instead of a shared one.

Reply:
I thought harder about this and this is the best I could come up with:

Manual bookkeeping is error prone. The only place that tracks these shared pointers is in `start_processor` and `entry_t`. Keeping the shared pointer in `entry_t` instead of `processor` means that the manager controls all the probes and it should be simpler to understand and harder to mess up.

`start_processor` is the harder one. This method cannot be called concurrently as is (there is a test that catches this, which is why we used to create probes in a different place and pass it around), because you could create multiple probes. I think for now this is the easier thing to reason about while the queue prevents concurrent code from running.

Reply:
sgtm. after i left this comment i ran into the later commit that addressed duplicate probe names and then mentally assumed my previous comment was out of date but forgot to update it.

Reply:
Yeah still valid, it's a little sneaky to reason about the inserts here as soon as you introduce any concurrency, so it's better to keep it simple for now.